logstash-core/lib/logstash/util.rb

```diff
@@ -183,4 +183,21 @@ def self.stringify_symbols(o)
       o
     end
   end
+
+  def self.deep_clone(o)
+    case o
+    when Hash
+      o.inject({}) {|h, (k,v)| h[k] = deep_clone(v); h }
+    when Array
+      o.map {|v| deep_clone(v) }
+    when Fixnum, Symbol, IO, TrueClass, FalseClass, NilClass
+      o
+    when LogStash::Codecs::Base
+      o.clone
```

Comment:
Codecs::Base does this
```ruby
  def clone
    return self.class.new(params)
  end
```
and this will call deep_clone of the params given.

BTW codec base is the only subclass that has a clone method.

If anything we should call register after clone
e.g. 
```ruby
  def self.deep_clone(o)
    case o
    when Hash
      o.inject({}) {|h, (k,v)| h[k] = deep_clone(v); h }
    when Array
      o.map {|v| deep_clone(v) }
    when Fixnum, Symbol, IO, TrueClass, FalseClass, NilClass
      o
    when LogStash::Codecs::Base
      o.clone.tap {|c| c.register }
    when String
      o.dup
    else
      Marshal.load(Marshal.dump(o))
    end
  end
```